### PR TITLE
Paging in shell for autocompletion and search

### PR DESCRIPTION
**Pull request by frazze-jobb**
Adds scrollable pager for large outputs for shell hotkeys.

### CODE REVIEW

**Comment by dgud** [comment]
I'm old school and don't want paging, can an option (or function) be added to turn off all paging?

**Comment by garazdawi** [comment]
Tests and docs?

**Review by garazdawi — CHANGES_REQUESTED** [review]
Neat! Some tests and this should be good to go.

**Comment by frazze-jobb** [comment]
> I'm old school and don't want paging, can an option (or function) be added to turn off all paging?

I'll do that in a separate PR, you can already disable paging for expand, but for c:h/* and c:i/* I wanted to try to make it use the new paging mechanism, but I didn't figure out how I could do that without deadlocking so I need to think some more.

File: lib/kernel/src/prim_tty.erl [hunk 1]
```diff
@@ -611,17 +612,53 @@ handle_request(State = #state{unicode = U, cols = W}, redraw_prompt_pre_deleted)
...
                             _ -> cols(State#state.buffer_before, U) + cols(State#state.buffer_after,U)
                           end,
-                          {ExpandBuffer, NewState} = insert_buf(RedrawState#state{ buffer_expand = [] }, iolist_to_binary(BufferExpand)),
+                          ERow = State#state.buffer_expand_row,
+
+                            BufferExpandLines = string:split(erlang:binary_to_list(BufferExpand), "\n", all),
+                            InputRows = (cols_multiline([State#state.buffer_before ++ State#state.buffer_after], W, U) div W),
+                            ExpandRows = (cols_multiline(BufferExpandLines, W, U) div W),
+                            BufferExpand1 = case ExpandRows > (R-InputRows) of
+                                true -> StatusLine = io_lib:format("\e[37;46mrows ~w to ~w of ~w\e[0m", [ERow, ERow + R-1-InputRows, ExpandRows]),
```
[the remark this answers is not in the document]
Comment (frazze-jobb) [reply lib/kernel/src/prim_tty.erl]: I made a function instead

File: lib/kernel/src/prim_tty.erl [hunk 2]
```diff
@@ -611,29 +640,77 @@ handle_request(State = #state{unicode = U, cols = W}, redraw_prompt_pre_deleted)
                             true when Last =/= [] -> cols(Last, U);
                             _ -> cols(State#state.buffer_before, U) + cols(State#state.buffer_after,U)
                           end,
-                          {ExpandBuffer, NewState} = insert_buf(RedrawState#state{ buffer_expand = [] }, iolist_to_binary(BufferExpand)),
+                          ERow = State#state.buffer_expand_row,
+
+                            BufferExpandLines = string:split(erlang:binary_to_list(BufferExpand), "\n", all),
```
Comment (garazdawi) [inline lib/kernel/src/prim_tty.erl:645]:
```suggestion
                            BufferExpandLines = string:split(unicode:characters_to_list(BufferExpand), "\n", all),
```

File: lib/kernel/src/prim_tty.erl [hunk 3]
```diff
@@ -611,29 +640,77 @@ handle_request(State = #state{unicode = U, cols = W}, redraw_prompt_pre_deleted)
...
+                                        {_, BufferExpandLinesInViewRev, _, {BEIVHalf, _}} = split_cols_multiline(Cols1, BufferExpandLinesInViewStart++[BEStartIVHalf], U, W),
+                                        BEIVHalf1 = case BEIVHalf of [] -> [];
+                                            _ -> [BEIVHalf]
+                                        end,
+                                        ExpandInView = lists:reverse(BEIVHalf1++BufferExpandLinesInViewRev),
+                                        ["\r\n",lists:join("\n", ExpandInView ++ [StatusLine])];
+                                false ->
+                                    ["\r\n",BufferExpand]
+                                end,
+                          {ExpandBuffer, NewState} = insert_buf(RedrawState#state{ buffer_expand = [] }, iolist_to_binary(BufferExpand1)),
```
Comment (garazdawi) [inline lib/kernel/src/prim_tty.erl:672]:
```suggestion
                          {ExpandBuffer, NewState} = insert_buf(RedrawState#state{ buffer_expand = [] }, unicode:characters_to_binary(BufferExpand1)),
```

File: lib/kernel/src/prim_tty.erl [hunk 4]
```diff
@@ -611,29 +640,77 @@ handle_request(State = #state{unicode = U, cols = W}, redraw_prompt_pre_deleted)
...
+                            InputRows = (cols_multiline([State#state.buffer_before ++ State#state.buffer_after], W, U) div W),
+                            ExpandRows = (cols_multiline(BufferExpandLines, W, U) div W),
+                            ExpandRowsLimit = case State#state.buffer_expand_limit of
+                                0 ->
+                                     ExpandRows;
+                                Limit ->
+                                    min(Limit, ExpandRows)
+                            end,
+                            ExpandRowsLimit1 = min(ExpandRowsLimit, R-1-InputRows),
+                            BufferExpand1 = case ExpandRows > (R-InputRows) of
```
Comment (garazdawi) [inline lib/kernel/src/prim_tty.erl:655]:
```suggestion
                            BufferExpand1 = case ExpandRows > ExpandRowsLimit1 of
```

File: lib/kernel/src/prim_tty.erl [hunk 5]
```diff
@@ -611,29 +640,77 @@ handle_request(State = #state{unicode = U, cols = W}, redraw_prompt_pre_deleted)
...
     {Output, State};
+handle_request(State = #state{ buffer_expand = Expand, buffer_expand_row = ERow, cols = W, rows = WindowRows, unicode = U}, {move_expand, N}) ->
+    %% Get number of Lines in terminal window
+    BufferExpandLines = case Expand of
+        undefined -> [];
+        _ -> string:split(erlang:binary_to_list(Expand), "\n", all)
+    end,
+    ExpandRows = (cols_multiline(BufferExpandLines, W, U) div W),
+    InputRows = (cols_multiline([State#state.buffer_before ++ State#state.buffer_after], W, U) div W),
+    ERow1 = if ExpandRows > WindowRows-InputRows -> %% We need to page expand rows
```
Comment (garazdawi) [inline lib/kernel/src/prim_tty.erl:686]: Something is off with this calculation. When the number of rows is less than windows size but larger than 7 the navigation does not work.

To see the effect type `l` `<TAB>`  and then try to navigate.

**Review by garazdawi — CHANGES_REQUESTED** [review]
If I type: `s` `<TAB>` `<END>` then any up/down/pageup/pagedown i type is ignored. Seems like maybe pressing `<END>`  should trigger autocompletion to be cancelled? When I precc `<HOME>` it is cancelled.

File: lib/stdlib/src/edlin.erl [hunk 6]
```diff
@@ -198,7 +206,8 @@ edit(Buf, P, {LB, {Bef,Aft}, LA}=MultiLine, {ShellMode, EscapePrefix}, Rs0) ->
                      reverse(Rs0)};
                 tab_expand_quit ->
                     %% When exiting tab expand mode, we want to evaluate the key in normal mode
-                    edit(Buf, P, MultiLine, {normal,none}, Rs0);
+                    %% we send a {move_rel, 0} event to make sure the paging area is cleared
+                    edit(Buf, P, MultiLine, {normal,none}, [{move_rel, 0}|Rs0]);
```
Comment (frazze-jobb) [inline lib/stdlib/src/edlin.erl:210]: @garazdawi: End will now close the expand, because of this added {move_rel, 0} event.

**Pull request merged**